File: skills/openai-image-gen/scripts/openai-image-gen.py
#!/usr/bin/env python3
"""openai-image-gen: text-to-image (gen) / image-to-image (edit) via any
OpenAI-compatible /v1/images gateway (new-api / sub2api / OpenAI direct).
Stdlib only, Python >= 3.9. Config priority: CLI flags > env vars > <skill-dir>/.env"""
import argparse
import base64
import json
import os
import re
import sys
import urllib.error
import urllib.request
from datetime import datetime
from pathlib import Path

SKILL_DIR = Path(__file__).resolve().parent.parent
ENV_FILE = SKILL_DIR / ".env"
MIME = {".png": "image/png", ".jpg": "image/jpeg", ".jpeg": "image/jpeg", ".webp": "image/webp", ".gif": "image/gif"}


def die(msg: str) -> None:
    print(msg, file=sys.stderr)
    sys.exit(1)


def load_env_file(path: Path) -> dict:
    out = {}
    if not path.exists():
        return out
    for raw in path.read_text(encoding="utf-8").splitlines():
        line = raw.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        k, v = line.split("=", 1)
        out[k.strip()] = v.strip().strip('"').strip("'")
    return out


ENV = load_env_file(ENV_FILE)


def pick(cli_val, name: str, default: str = "") -> str:
    return cli_val or os.environ.get(name) or ENV.get(name) or default


def to_image_ref(spec: str, label: str) -> dict:
    # local file -> data URL (OpenAI edits JSON accepts image_url only)
    if re.match(r"^(https?|data):", spec, re.I):
        return {"image_url": spec}
    p = Path(spec).resolve()
    if not p.is_file():
        die(f"{label} file not found: {p}")
    mime = MIME.get(p.suffix.lower())
    if not mime:
        die(f"{label} unsupported extension (png/jpg/jpeg/webp/gif): {p}")
    return {"image_url": f"data:{mime};base64,{base64.b64encode(p.read_bytes()).decode()}"}


def post(url: str, key: str, payload: dict):
    body = json.dumps(payload, ensure_ascii=False).encode("utf-8")
    req = urllib.request.Request(
        url, data=body, method="POST",
        headers={"authorization": f"Bearer {key}", "content-type": "application/json"},
    )
    try:
        # no timeout: high-quality large images can take minutes, bounded by agent-side terminal timeout
        with urllib.request.urlopen(req) as r:
            return r.status, r.read().decode("utf-8", "replace")
    except urllib.error.HTTPError as e:
        return e.code, e.read().decode("utf-8", "replace")
    except Exception as e:
        die(f"request failed: {e} (check BASE_URL or network/proxy)")


def build_parser() -> argparse.ArgumentParser:
    ap = argparse.ArgumentParser(prog="openai-image-gen", description="Text-to-image / image-to-image via any OpenAI-compatible /v1/images gateway")
    sub = ap.add_subparsers(dest="cmd", required=True)
    for name in ("gen", "edit"):
        sp = sub.add_parser(name)
        sp.add_argument("--prompt", required=True)
        sp.add_argument("--model")
        sp.add_argument("--size")
        sp.add_argument("--quality")
        sp.add_argument("--n", type=int)
        sp.add_argument("--out")
        sp.add_argument("--out-dir", dest="out_dir")
        sp.add_argument("--base-url", dest="base_url")
        sp.add_argument("--api-key", dest="api_key")
        if name == "edit":
            sp.add_argument("--image", action="append", required=True, help="reference image: local file or http(s) URL, repeatable")
            sp.add_argument("--mask", help="mask image: local file or http(s) URL")
    return ap


def main() -> None:
    a = build_parser().parse_args()

    base = pick(a.base_url, "IMGAPI_BASE_URL").rstrip("/")
    key = pick(a.api_key, "IMGAPI_API_KEY")
    model = pick(a.model, "IMGAPI_MODEL", "gpt-image-2")
    if not base or not key:
        die(f"IMGAPI_BASE_URL / IMGAPI_API_KEY missing: set --base-url/--api-key or env vars, or write {ENV_FILE}")

    endpoint = "/v1/images/generations" if a.cmd == "gen" else "/v1/images/edits"
    payload = {"model": model, "prompt": a.prompt}
    if a.size:
        payload["size"] = a.size
    if a.quality:
        payload["quality"] = a.quality
    if a.n:
        payload["n"] = a.n
    if a.cmd == "edit":
        payload["images"] = [to_image_ref(s, "--image") for s in a.image]
        if a.mask:
            payload["mask"] = to_image_ref(a.mask, "--mask")

    status, text = post(base + endpoint, key, payload)
    if status >= 400:
        die(f"HTTP {status}\n{text[:2000]}")
    try:
        data = json.loads(text)["data"]
    except Exception:
        die(f"non-JSON response:\n{text[:500]}")
    if not data:
        die("response data[] is empty")
    if a.out and len(data) > 1:
        die("--out fits a single image only; use --out-dir for multiple")

    now = datetime.now().strftime("%Y%m%d-%H%M%S")
    out_path = Path(a.out).resolve() if a.out else None
    out_dir = out_path.parent if out_path else Path(a.out_dir or Path.cwd()).resolve()
    out_dir.mkdir(parents=True, exist_ok=True)
    prefix = "t2i" if a.cmd == "gen" else "i2i"

    for i, item in enumerate(data, 1):
        m = re.search(r"\.(png|jpe?g|webp|gif)(?:\?|$)", str(item.get("url", "")), re.I)
        ext = m.group(1).lower().replace("jpeg", "jpg") if m else "png"
        f = out_path or out_dir / f"{prefix}-{now}{'' if len(data) == 1 else f'-{i}'}.{ext}"
        if item.get("b64_json"):
            buf = base64.b64decode(item["b64_json"])
        elif item.get("url"):
            try:
                with urllib.request.urlopen(item["url"], timeout=120) as r:
                    buf = r.read()
            except Exception as e:
                die(f"failed to download result image: {e} url={str(item['url'])[:120]}")
        else:
            die("response contains neither b64_json nor url")
        f.write_bytes(buf)
        print(f"SAVED {f} ({len(buf)} bytes, model={model})")


if __name__ == "__main__":
    main()

File: skills/openai-image-gen/scripts/openai-image-gen.mjs
#!/usr/bin/env node
// openai-image-gen: text-to-image (gen) / image-to-image (edit) via any
// OpenAI-compatible /v1/images gateway (new-api / sub2api / OpenAI direct).
// Zero dependency, Node >= 18. Config priority: CLI flags > process env > <skill-dir>/.env
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const SKILL_DIR = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const ENV_FILE = path.join(SKILL_DIR, '.env');
const USAGE = `usage:
  node openai-image-gen.mjs gen  --prompt "..." [--model m] [--size 1024x1024] [--quality auto|low|medium|high] [--n 1] [--out file.png | --out-dir DIR]
  node openai-image-gen.mjs edit --prompt "..." --image <path|url> [--image ...] [--mask <path|url>] [same model/size/out flags]
config: --base-url/--api-key flags, env IMGAPI_BASE_URL/IMGAPI_API_KEY, or ${ENV_FILE}`;

const die = (msg) => {
  console.error(msg);
  process.exit(1);
};

const KNOWN_FLAGS = ['prompt', 'model', 'size', 'quality', 'n', 'out', 'out-dir', 'base-url', 'api-key', 'mask', 'image'];
const MIME = { '.png': 'image/png', '.jpg': 'image/jpeg', '.jpeg': 'image/jpeg', '.webp': 'image/webp', '.gif': 'image/gif' };

function parseEnvFile(text) {
  const out = {};
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.trim();
    if (!line || line.startsWith('#') || !line.includes('=')) continue;
    const i = line.indexOf('=');
    out[line.slice(0, i).trim()] = line.slice(i + 1).trim().replace(/^"(.*)"$/, '$1');
  }
  return out;
}

// local file -> data URL (OpenAI edits JSON accepts image_url only, no file upload needed)
function toImageRef(spec, label) {
  if (/^(https?:|data:)/i.test(spec)) return { image_url: spec };
  const abs = path.resolve(spec);
  if (!fs.existsSync(abs)) die(`${label} file not found: ${abs}`);
  const mime = MIME[path.extname(abs).toLowerCase()];
  if (!mime) die(`${label} unsupported extension (png/jpg/jpeg/webp/gif): ${abs}`);
  return { image_url: `data:${mime};base64,${fs.readFileSync(abs).toString('base64')}` };
}

async function main() {
  const argv = process.argv.slice(2);
  const cmd = argv.shift();
  if (!['gen', 'edit'].includes(cmd)) die(`unknown command: "${cmd ?? ''}"\n${USAGE}`);
  const flags = { image: [] };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (!a.startsWith('--')) die(`unexpected argument: ${a}\n${USAGE}`);
    const key = a.slice(2);
    if (!KNOWN_FLAGS.includes(key)) die(`unknown flag --${key}\n${USAGE}`);
    const val = argv[++i];
    if (val === undefined) die(`missing value for --${key}`);
    if (key === 'image') flags.image.push(val);
    else flags[key] = val;
  }
  if (!flags.prompt) die(`missing --prompt\n${USAGE}`);
  if (flags.n !== undefined && (!Number.isInteger(Number(flags.n)) || Number(flags.n) <= 0)) die('--n must be a positive integer');

  const env = fs.existsSync(ENV_FILE) ? parseEnvFile(fs.readFileSync(ENV_FILE, 'utf8')) : {};
  const baseUrl = (flags['base-url'] || process.env.IMGAPI_BASE_URL || env.IMGAPI_BASE_URL || '').replace(/\/+$/, '');
  const apiKey = flags['api-key'] || process.env.IMGAPI_API_KEY || env.IMGAPI_API_KEY || '';
  const model = flags.model || process.env.IMGAPI_MODEL || env.IMGAPI_MODEL || 'gpt-image-2';
  if (!baseUrl || !apiKey) die(`IMGAPI_BASE_URL / IMGAPI_API_KEY missing: pass --base-url/--api-key, set env vars, or write ${ENV_FILE}`);

  const endpoint = cmd === 'gen' ? '/v1/images/generations' : '/v1/images/edits';
  const payload = { model, prompt: flags.prompt };
  if (flags.size) payload.size = flags.size;
  if (flags.quality) payload.quality = flags.quality;
  if (flags.n) payload.n = Number(flags.n);
  if (cmd === 'edit') {
    if (!flags.image.length) die('edit requires at least one --image (local path or http(s) URL)');
    payload.images = flags.image.map((s) => toImageRef(s, '--image'));
    if (flags.mask) payload.mask = toImageRef(flags.mask, '--mask');
  }

  let res;
  let text;
  try {
    // no fetch-level timeout: high-quality large images can take minutes, bounded by agent-side terminal timeout
    res = await fetch(baseUrl + endpoint, {
      method: 'POST',
      headers: { authorization: `Bearer ${apiKey}`, 'content-type': 'application/json' },
      body: JSON.stringify(payload),
    });
    text = await res.text();
  } catch (e) {
    die(`request failed: ${e.message} (check BASE_URL or network/proxy)`);
  }
  if (!res.ok) die(`HTTP ${res.status}\n${text.slice(0, 2000)}`);
  let data;
  try {
    data = JSON.parse(text).data;
  } catch {
    die(`non-JSON response:\n${text.slice(0, 500)}`);
  }
  if (!Array.isArray(data) || data.length === 0) die('response data[] is empty');
  if (flags.out && data.length > 1) die('--out fits a single image only; use --out-dir for multiple');

  const d = new Date();
  const p2 = (x) => String(x).padStart(2, '0');
  const stamp = `${d.getFullYear()}${p2(d.getMonth() + 1)}${p2(d.getDate())}-${p2(d.getHours())}${p2(d.getMinutes())}${p2(d.getSeconds())}`;
  const outFile = flags.out ? path.resolve(flags.out) : null;
  const outDir = outFile ? path.dirname(outFile) : path.resolve(flags['out-dir'] || process.cwd());
  fs.mkdirSync(outDir, { recursive: true });
  const prefix = cmd === 'gen' ? 't2i' : 'i2i';

  let idx = 0;
  for (const item of data) {
    idx += 1;
    let buf;
    if (item.b64_json) {
      buf = Buffer.from(item.b64_json, 'base64');
    } else if (item.url) {
      const dl = await fetch(item.url).catch((e) => die(`failed to download result image: ${e.message} url=${item.url.slice(0, 120)}`));
      if (!dl.ok) die(`failed to download result image: HTTP ${dl.status} url=${item.url.slice(0, 120)}`);
      buf = Buffer.from(await dl.arrayBuffer());
    } else {
      die('response contains neither b64_json nor url');
    }
    const m = item.url && item.url.match(/\.(png|jpe?g|webp|gif)(?:\?|$)/i);
    const ext = m ? m[1].toLowerCase().replace('jpeg', 'jpg') : 'png';
    const file = outFile || path.join(outDir, `${prefix}-${stamp}${data.length > 1 ? `-${idx}` : ''}.${ext}`);
    fs.writeFileSync(file, buf);
    console.log(`SAVED ${file} (${buf.length} bytes, model=${model})`);
  }
}

main().catch((e) => die(String((e && e.stack) || e)));
